File: alternating_bit_protocol.c
/* Deterministic contract lowered from the reviewed V2 domain 'alternating_bit_protocol_with_lossy_channels'.
 * Human review of the reviewed artifact is required before trust. */

typedef struct {
    int sender_bit;
    int receiver_bit;
    int msg_channel;
    int ack_channel;
} alternating_bit_protocol_with_lossy_channels;

/*@
  requires \valid(counter);
  assigns counter->sender_bit, counter->receiver_bit, counter->msg_channel, counter->ack_channel;
  ensures counter->sender_bit == 0 && counter->receiver_bit == 0 && counter->msg_channel == -1 && counter->ack_channel == -1;
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_init(alternating_bit_protocol_with_lossy_channels *counter) {
    counter->sender_bit = 0;
    counter->receiver_bit = 0;
    counter->msg_channel = -1;
    counter->ack_channel = -1;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->sender_bit;
*/
int alternating_bit_protocol_with_lossy_channels_get_sender_bit(const alternating_bit_protocol_with_lossy_channels *counter) {
    return counter->sender_bit;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->receiver_bit;
*/
int alternating_bit_protocol_with_lossy_channels_get_receiver_bit(const alternating_bit_protocol_with_lossy_channels *counter) {
    return counter->receiver_bit;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->msg_channel;
*/
int alternating_bit_protocol_with_lossy_channels_get_msg_channel(const alternating_bit_protocol_with_lossy_channels *counter) {
    return counter->msg_channel;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->ack_channel;
*/
int alternating_bit_protocol_with_lossy_channels_get_ack_channel(const alternating_bit_protocol_with_lossy_channels *counter) {
    return counter->ack_channel;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
  requires counter->msg_channel == -1;
  assigns counter->msg_channel;
  ensures counter->msg_channel == \old(counter->sender_bit);
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_send_msg(alternating_bit_protocol_with_lossy_channels *counter) {
    int pre_sender_bit = counter->sender_bit;
    counter->msg_channel = pre_sender_bit;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
  requires counter->msg_channel != -1;
  assigns counter->msg_channel;
  ensures counter->msg_channel == -1;
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_drop_msg(alternating_bit_protocol_with_lossy_channels *counter) {
    counter->msg_channel = -1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
  requires (counter->msg_channel == counter->receiver_bit) && (counter->ack_channel == -1);
  assigns counter->ack_channel, counter->msg_channel, counter->receiver_bit;
  ensures counter->ack_channel == \old(counter->receiver_bit) && counter->msg_channel == -1 && counter->receiver_bit == 1 - \old(counter->receiver_bit);
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_receive_msg(alternating_bit_protocol_with_lossy_channels *counter) {
    int pre_receiver_bit = counter->receiver_bit;
    counter->ack_channel = pre_receiver_bit;
    counter->msg_channel = -1;
    counter->receiver_bit = 1 - pre_receiver_bit;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
  requires counter->ack_channel != -1;
  assigns counter->ack_channel;
  ensures counter->ack_channel == -1;
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_drop_ack(alternating_bit_protocol_with_lossy_channels *counter) {
    counter->ack_channel = -1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
  requires counter->ack_channel == counter->sender_bit;
  assigns counter->ack_channel, counter->sender_bit;
  ensures counter->ack_channel == -1 && counter->sender_bit == 1 - \old(counter->sender_bit);
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_receive_ack(alternating_bit_protocol_with_lossy_channels *counter) {
    int pre_sender_bit = counter->sender_bit;
    counter->ack_channel = -1;
    counter->sender_bit = 1 - pre_sender_bit;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
  requires ((counter->msg_channel != -1) && (counter->msg_channel != counter->receiver_bit)) && (counter->ack_channel == -1);
  assigns counter->ack_channel, counter->msg_channel;
  ensures counter->ack_channel == \old(counter->msg_channel) && counter->msg_channel == -1;
  ensures (0 <= counter->sender_bit) && (counter->sender_bit <= 1) && (0 <= counter->receiver_bit) && (counter->receiver_bit <= 1) && (-1 <= counter->msg_channel) && (counter->msg_channel <= 1) && (-1 <= counter->ack_channel) && (counter->ack_channel <= 1) && ((counter->msg_channel == counter->receiver_bit) ==> (counter->sender_bit == counter->receiver_bit)) && ((counter->ack_channel == counter->sender_bit) ==> (counter->sender_bit != counter->receiver_bit)) && ((counter->sender_bit == counter->receiver_bit) ==> (counter->ack_channel != counter->sender_bit)) && ((counter->sender_bit != counter->receiver_bit) ==> (counter->msg_channel != counter->receiver_bit));
*/
void alternating_bit_protocol_with_lossy_channels_resend_ack(alternating_bit_protocol_with_lossy_channels *counter) {
    int pre_msg_channel = counter->msg_channel;
    counter->ack_channel = pre_msg_channel;
    counter->msg_channel = -1;
}

File: AlternatingBitProtocol.rs
use prusti_contracts::*;

pub struct AlternatingBitProtocolWithLossyChannels {
    pub sender_bit: i32,
    pub receiver_bit: i32,
    pub msg_channel: i32,
    pub ack_channel: i32,
}

impl AlternatingBitProtocolWithLossyChannels {
    #[ensures(result.sender_bit == 0 && result.receiver_bit == 0 && result.msg_channel == -1 && result.ack_channel == -1 && (0 <= result.sender_bit) && (result.sender_bit <= 1) && (0 <= result.receiver_bit) && (result.receiver_bit <= 1) && (-1 <= result.msg_channel) && (result.msg_channel <= 1) && (-1 <= result.ack_channel) && (result.ack_channel <= 1) && ((result.msg_channel == result.receiver_bit) ==> (result.sender_bit == result.receiver_bit)) && ((result.ack_channel == result.sender_bit) ==> (result.sender_bit != result.receiver_bit)) && ((result.sender_bit == result.receiver_bit) ==> (result.ack_channel != result.sender_bit)) && ((result.sender_bit != result.receiver_bit) ==> (result.msg_channel != result.receiver_bit)))]
    pub fn new() -> Self {
        Self { sender_bit: 0, receiver_bit: 0, msg_channel: -1, ack_channel: -1 }
    }

    #[pure]
    #[ensures(result == self.sender_bit)]
    pub fn get_sender_bit(&self) -> i32 {
        self.sender_bit
    }

    #[pure]
    #[ensures(result == self.receiver_bit)]
    pub fn get_receiver_bit(&self) -> i32 {
        self.receiver_bit
    }

    #[pure]
    #[ensures(result == self.msg_channel)]
    pub fn get_msg_channel(&self) -> i32 {
        self.msg_channel
    }

    #[pure]
    #[ensures(result == self.ack_channel)]
    pub fn get_ack_channel(&self) -> i32 {
        self.ack_channel
    }

    #[requires((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    #[requires(self.msg_channel == -1)]
    #[ensures(self.msg_channel == old(self.sender_bit))]
    #[ensures((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    pub fn send_msg(&mut self) {
        let pre_sender_bit = self.sender_bit;
        self.msg_channel = pre_sender_bit;
    }

    #[requires((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    #[requires(self.msg_channel != -1)]
    #[ensures(self.msg_channel == -1)]
    #[ensures((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    pub fn drop_msg(&mut self) {
        self.msg_channel = -1;
    }

    #[requires((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    #[requires((self.msg_channel == self.receiver_bit) && (self.ack_channel == -1))]
    #[ensures(self.ack_channel == old(self.receiver_bit) && self.msg_channel == -1 && self.receiver_bit == 1 - old(self.receiver_bit))]
    #[ensures((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    pub fn receive_msg(&mut self) {
        let pre_receiver_bit = self.receiver_bit;
        self.ack_channel = pre_receiver_bit;
        self.msg_channel = -1;
        self.receiver_bit = 1 - pre_receiver_bit;
    }

    #[requires((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    #[requires(self.ack_channel != -1)]
    #[ensures(self.ack_channel == -1)]
    #[ensures((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    pub fn drop_ack(&mut self) {
        self.ack_channel = -1;
    }

    #[requires((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    #[requires(self.ack_channel == self.sender_bit)]
    #[ensures(self.ack_channel == -1 && self.sender_bit == 1 - old(self.sender_bit))]
    #[ensures((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    pub fn receive_ack(&mut self) {
        let pre_sender_bit = self.sender_bit;
        self.ack_channel = -1;
        self.sender_bit = 1 - pre_sender_bit;
    }

    #[requires((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    #[requires(((self.msg_channel != -1) && (self.msg_channel != self.receiver_bit)) && (self.ack_channel == -1))]
    #[ensures(self.ack_channel == old(self.msg_channel) && self.msg_channel == -1)]
    #[ensures((0 <= self.sender_bit) && (self.sender_bit <= 1) && (0 <= self.receiver_bit) && (self.receiver_bit <= 1) && (-1 <= self.msg_channel) && (self.msg_channel <= 1) && (-1 <= self.ack_channel) && (self.ack_channel <= 1) && ((self.msg_channel == self.receiver_bit) ==> (self.sender_bit == self.receiver_bit)) && ((self.ack_channel == self.sender_bit) ==> (self.sender_bit != self.receiver_bit)) && ((self.sender_bit == self.receiver_bit) ==> (self.ack_channel != self.sender_bit)) && ((self.sender_bit != self.receiver_bit) ==> (self.msg_channel != self.receiver_bit)))]
    pub fn resend_ack(&mut self) {
        let pre_msg_channel = self.msg_channel;
        self.ack_channel = pre_msg_channel;
        self.msg_channel = -1;
    }
}
